atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraMutationTimestampProvider.java

```diff
@@ -16,34 +16,47 @@
 
 package com.palantir.atlasdb.cassandra;
 
-import com.google.common.collect.Multimap;
-import com.palantir.atlasdb.keyvalue.api.TableReference;
+import java.util.function.LongUnaryOperator;
 
 /**
  * This interface produces Cassandra timestamps for mutations, where it makes sense for these to be overridden.
- * Generally, to ensure that Cassandra is able to compact tombstones smoothly, these values should increase over
- * time and should largely be in line with AtlasDB's notion of logical time.
+ * The values returned by this provider must observe the following semantics.
  *
- * Calling these methods must be cheap, at least on an amortised basis; they may be called multiple times
- * in rapid succession (as frequently as once per cell) when
- * {@link com.palantir.atlasdb.keyvalue.api.KeyValueService#delete(TableReference, Multimap)} or other deletion
- * methods are called.
+ *   - Sweep sentinels: These must be strictly greater than the Cassandra timestamps of any previous deletions or range
+ *     tombstones covering the sweep sentinel. (We may want to write sweep sentinels to a table where deletions or range
+ *     tombstones already exist, in the case of tables shifting between THOROUGH to CONSERVATIVE sweep.)
+ *   - Deletions: These must be strictly greater than the Cassandra timestamp at which the original value being deleted
+ *     was written.
+ *   - Range tombstones: These must be strictly greater than the Cassandra timestamps of all values covered by the
+ *     range tombstone (including sweep sentinels, if applicable.)
+ *
+ * In addition to the above semantics, these values should increase over time and should largely be in line with
+ * AtlasDB's notion of logical time, to ensure smooth compaction of tables.
  */
 public interface CassandraMutationTimestampProvider {
     /**
-     * Cassandra timestamp at which sweep sentinels should be written.
+     * Returns the Cassandra timestamp at which sweep sentinels should be written.
+     *
+     * This is called once per API call to CassandraKeyValueServiceImpl.addGarbageCollectionSentinelValues().
      */
     long getSweepSentinelWriteTimestamp();
 
     /**
-     * Cassandra timestamp at which KVS-level deletes should be performed.
-     * @param atlasDeletionTimestamp Atlas timestamp of the relevant delete.
+     * Returns an operator, which determines the Cassandra timestamp to write a delete at, given the Atlas timestamp
+     * at which the deletion occurred.
+     * The operator is intended to be use for a single batch deletion, and must not be re-used outside of the
+     * context of a single batch delete.
+     *
+     * The operator is obtained once per API call to CassandraKeyValueServiceImpl.delete().
+     * The operator is invoked once per cell involved in such an API call, and thus must be cheap to invoke.
```

Comment:
Is this still correct?

Reply:
Yep - see `CellDeleter:96`. The problem here is that in general you can't call this outside the loop, because in legacy/test mode you want to return `ts + 1` where ts is the delete timestamp